moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+        i++
+      }
+      return true
+    }
+
+    /**
+     * Returns an object that calls a `method` method on `adapter` in service of
+     * converting an object from JSON.
+     */
+    fun fromAdapter(adapter: Any, method: Method): AdapterMethod {
+      method.isAccessible = true
+      val returnType = method.genericReturnType
+      val returnTypeAnnotations = method.jsonAnnotations
+      val parameterTypes = method.genericParameterTypes
+      val parameterAnnotations = method.parameterAnnotations
+      return if (parameterTypes.isNotEmpty() &&
+        parameterTypes[0] == JsonReader::class.java &&
+        returnType != Void.TYPE &&
+        parametersAreJsonAdapters(1, parameterTypes)
+      ) {
+        // Point pointFromJson(JsonReader jsonReader) {
+        // Point pointFromJson(JsonReader jsonReader, JsonAdapter<?> adapter, ...) {
+        object : AdapterMethod(
+          returnType, returnTypeAnnotations, adapter, method, parameterTypes.size, 1, true
+        ) {
+          override fun fromJson(moshi: Moshi, reader: JsonReader): Any? {
+            return invoke(reader)
+          }
+        }
+      } else if (parameterTypes.size == 1 && returnType != Void.TYPE) {
+        // Point pointFromJson(List<Integer> o) {
+        val qualifierAnnotations = parameterAnnotations[0].jsonAnnotations
+        val nullable = parameterAnnotations[0].hasNullable
+        object : AdapterMethod(
+          returnType, returnTypeAnnotations, adapter, method, parameterTypes.size, 1, nullable
+        ) {
+          var delegate: JsonAdapter<Any>? = null
+
+          override fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+            super.bind(moshi, factory)
+            delegate = if (Types.equals(parameterTypes[0], returnType) &&
+              qualifierAnnotations == returnTypeAnnotations
+            ) moshi.nextAdapter(factory, parameterTypes[0], qualifierAnnotations)
+            else moshi.adapter(parameterTypes[0], qualifierAnnotations)
+          }
+
+          override fun fromJson(moshi: Moshi, reader: JsonReader): Any? {
+            val intermediate = delegate!!.fromJson(reader)
+            return invoke(intermediate)
+          }
+        }
+      } else {
+        throw IllegalArgumentException(
+          """Unexpected signature for $method.
+@FromJson method signatures may have one of the following structures:
+    <any access modifier> R fromJson(JsonReader jsonReader) throws <any>;
+    <any access modifier> R fromJson(JsonReader jsonReader, JsonAdapter<any> delegate, <any more delegates>) throws <any>;
+    <any access modifier> R fromJson(T value) throws <any>;
+"""
+        )
+      }
+    }
+
+    /** Returns the matching adapter method from the list.  */
+    private fun get(
+      adapterMethods: List<AdapterMethod>,
+      type: Type,
+      annotations: Set<Annotation>
+    ): AdapterMethod? {
+      var i = 0
+      val size = adapterMethods.size
+      while (i < size) {
```

Comment:
for-loop?

Reply:
https://github.com/square/moshi/pull/1467/commits/3e6e9bda3372997dff18ac5d1567f6b80707074e#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R302